### PR TITLE
tiledb: use array metadata to store xml

### PR DESCRIPTION
**Pull request by normanb**
##  What does this PR do?

This PR uses the metadata feature that is available in TileDB 1.7+ rather than creating a sidecar file. The benefits are API compatibility with TileDB as well as supporting versioned metadata. Backwards compatibility with the sidecar approach has been maintained where possible. 

## Tasklist

 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

File: gdal/frmts/tiledb/tiledbdataset.cpp [hunk 1]
```diff
@@ -696,24 +722,51 @@ CPLErr TileDBDataset::TryLoadCachedXML( char ** /*papszSiblingFiles*/, bool bRel
         {
             CPLErrorHandlerPusher oQuietError( CPLQuietErrorHandler );
 
-            if ( vfs.is_file( psPam->pszPamFilename ) )
+#if TILEDB_VERSION_MAJOR >= 1 && TILEDB_VERSION_MINOR >= 7
+            if ( bReload )
             {
-                if ( bReload )
+                tiledb_datatype_t v_type = TILEDB_UINT8; // CPLSerializeXMLTree returns char*
+                const void* v_r;
```
Comment (rouault) [inline gdal/frmts/tiledb/tiledbdataset.cpp]: perhaps initialize v_r and v_num so we are sure they are always at well defined values ?

File: gdal/frmts/tiledb/tiledbdataset.cpp [hunk 2]
```diff
@@ -696,24 +722,51 @@ CPLErr TileDBDataset::TryLoadCachedXML( char ** /*papszSiblingFiles*/, bool bRel
...
-                    std::istream is ( &fbuf );
-                    osMetaDoc.resize(nBytes);
-                    is.read( ( char* ) osMetaDoc.data(), nBytes );
-                    fbuf.close();
-                    psTree = CPLParseXMLString( osMetaDoc );
+                    auto oMeta = std::unique_ptr<tiledb::Array>( 
+                        new tiledb::Array( *m_ctx, m_array->uri(), TILEDB_READ )
+                    );
+                    oMeta->get_metadata("_gdal", &v_type, &v_num, &v_r);
+                    if ( v_r != NULL )
```
Comment (mloskot) [inline gdal/frmts/tiledb/tiledbdataset.cpp]: `if ( v_r != nullptr )` or `if ( v_r )`

File: gdal/frmts/tiledb/tiledbdataset.cpp [hunk 3]
```diff
@@ -696,24 +722,51 @@ CPLErr TileDBDataset::TryLoadCachedXML( char ** /*papszSiblingFiles*/, bool bRel
...
+                    {
+                        osMetaDoc = static_cast<const char*>( v_r );
+                    }
                 }
                 else
                 {
-                    psTree = CPLParseXMLString( osMetaDoc );
+                    m_array->get_metadata("_gdal", &v_type, &v_num, &v_r);
+                    if ( v_r != NULL )
```
Comment (mloskot) [inline gdal/frmts/tiledb/tiledbdataset.cpp]: `if ( v_r != nullptr )`

File: gdal/frmts/tiledb/tiledbdataset.cpp (hunk 1 above)
Comment (normanb) [reply gdal/frmts/tiledb/tiledbdataset.cpp]: +1

File: gdal/frmts/tiledb/tiledbdataset.cpp (hunk 2 above)
Comment (normanb) [reply gdal/frmts/tiledb/tiledbdataset.cpp]: +1

File: gdal/frmts/tiledb/tiledbdataset.cpp (hunk 3 above)
Comment (normanb) [reply gdal/frmts/tiledb/tiledbdataset.cpp]: +1

File: gdal/frmts/tiledb/tiledbdataset.cpp [hunk 4]
```diff
@@ -696,24 +724,51 @@ CPLErr TileDBDataset::TryLoadCachedXML( char ** /*papszSiblingFiles*/, bool bRel
         {
             CPLErrorHandlerPusher oQuietError( CPLQuietErrorHandler );
 
-            if ( vfs.is_file( psPam->pszPamFilename ) )
+#if TILEDB_VERSION_MAJOR >= 1 && TILEDB_VERSION_MINOR >= 7
+            if ( bReload )
             {
-                if ( bReload )
+                tiledb_datatype_t v_type = TILEDB_UINT8; // CPLSerializeXMLTree returns char*
+                const void* v_r = NULL;
```
Comment (rouault) [inline gdal/frmts/tiledb/tiledbdataset.cpp]:
```suggestion
                const void* v_r = nullptr;
```

**Pull request merged**